doc/designs/application-specific-passwords.md

```diff
@@ -0,0 +1,243 @@
+
+By default, the password will have no permissions assigned, being inactive in
+practice, with the user then explicitly assigning permissions for it at his/her
+discretion. This may range from access to only one particular service up to
+either (TBD):
+* access rights on par with his primary password, or
+* access to all systems and services, but not to the account management.
+
+It would be better for this to be at least to some extent separate from the
+general access control, as permissions for passwords will be set by the user,
+while permissions for the whole account are often set by other people. As a part
+of this, it may be desirable to implement an attribute determining whether
+default and managed permissions apply for the given user's password, or not.
+
+There is another option: in theory, authentication indicators could be used for
+access management for multiple passwords for a single user - it would then be
+best to differentiate 'levels' of passwords. This might (depending on
+implementation) facilitate less fine-grained access control for varius
+passwords, but would probably be easier to implement.
+
+If needed, self-service ACIs and/or delegation ACIs will be extended as well
+to support application-specific passwords.
+
+#### HBAC for application-specific passwords
+
+There are two options how to make HBAC distinguish between various passwords:
+* Assign permissions for access to hosts as a part of the permissions management
+  for the user's passwords. HBAC would be extended to check these permissions.
+* Extend HBAC rules to *(host, service, user, password id/password level)*
+  so that they will directly accommodate this functionality. The attribute
+  identifying a user's password or a password level:
+  * can either be an optional field or it can have a special wildcard value denoting
+    'all user's passwords',
+  * will be administered by the user (in line with permissions management for
+    user's passwords in general).
+
+### ipa-kdb support for the aforementioned functionality
+
+The ipa-kdb daemon will be extended to support the aforementioned functionality
+as well, so that it can be used with Kerberos, which plays a critical role in
+FreeIPA, as all commands facilitating user management require not only relevant
+permissions, but also an obtained Kerberos ticket.
+
+### SSSD support for the aforementioned functionality.
+
+SSSD will also be extended to support the aforementioned functionality. This
+will be done by a closely related, but separate extension of SSSD. On the
+side of FreeIPA, after SSSD is extended, interactions with SSSD using the new
+functionality must be tested.
+
+## Implementation
+
+The extension must be backward compatible, that means not breaking any existing
+functionality and passing all existing tests.
+
+As for changes to the data storage:
+* As mentioned above, if the attribute for password in the user schema is not
+  multi-valued, it must be modified to be.
```

Comment:
No it should not, it will probably break external tools, and it also makes it very hard to figure what's what unless subattributes are used and even then it will not work with kerberos which uses different keys in another object.
This whole part need to be analyzed in much more detail.

Reply:
I proposed an alternative solution to this in the [answer above](https://github.com/freeipa/freeipa/pull/4061#discussion_r364703681). But from this comment I see that there is an alternative solution to that attribute being a dictionary: assigning labels as a subattribute. This may be easier, as I still don't know how would I specify the attribute as a dictionary with the key being the Str type (holding labels) and value being the Password type (holding the password in the same protected way a in the case of the `userpassword` attribute). I will analyze what of these alternatives will be better, along with other aspects of this, of course.